### PR TITLE
feat: Scope slowTestThreshold to specific testing types

### PR DESCRIPTION
**Pull request by BlueWinds**
- Closes https://cypress-io.atlassian.net/browse/UNIFY-1302

### User facing changelog
`slowTestThreshold` is no longer a valid global configuration option, and must be defined per testing type. This will be done as part of the automatic config migration where possible.

### Additional details

If a user in 9.x has defined a custom slowTestThreshold at the global level:
```
{
  slowTestThreshold: 500
}
```

After this change, migrate to 10.x will update their config to this instead:
```
{
  e2e: {
    slowTestThreshold: 500
  },
  component: {
    slowTestThreshold: 500
  }
}
```

If they only have one testing type set up (eg, most users with only e2e tests), they will get the shorter version with slowTestThreshold moved inside their testing type:

```
{
  e2e: {
    slowTestThreshold: 500
  }
}
```

### PR Tasks
- [x] Have tests been added/updated?
- [x] Has the original issue (or this PR, if no issue exists) been tagged with a release in ZenHub? (user-facing changes only)
- [ ] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? TODO
- [x] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
- [N/A] Have new configuration options been added to the [`cypress.schema.json`](https://github.com/cypress-io/cypress/blob/develop/cli/schema/cypress.schema.json)?

### CODE REVIEW

File: packages/data-context/test/unit/sources/migration/codegen.spec.ts [hunk 1]
```diff
@@ -25,7 +25,7 @@ describe('cypress.config.js generation', () => {
 
     const generatedConfig = await createConfigString(config, {
       hasE2ESpec: true,
-      hasComponentTesting: false,
+      hasComponentTesting: true,
```
Comment (elevatebart) [inline packages/data-context/test/unit/sources/migration/codegen.spec.ts:28]: Why was this change necessary? 
It feels like something that should generate a different string than before.

We also stopped testing `hasComponenttesting: false` altogether

**Review by elevatebart — CHANGES_REQUESTED** [review]
A few changes that I don't understand. 

Also, can you add a test for migration that actually moves the value from the root of the json file to the e2e tests?

File: packages/config/lib/options.ts [hunk 2]
```diff
@@ -560,6 +560,12 @@ export const breakingRootOptions: Array<BreakingOption> = [
     isWarning: false,
     testingTypes: ['e2e'],
   },
+  {
+    name: 'slowTestThreshold',
```
Comment (lmiller1990) [inline packages/config/lib/options.ts:564]: When was it decided this should not be a root level config? This seems inconsistent to me; although CT and E2E have different defaults for `viewportWidth` and `viewportHeight`, we don't stop the user from configuring those as root values. Why is `slowTestThreshold` different? If I want to have a single value for both, I think I should be allowed to.

**Review by lmiller1990 — COMMENTED** [review]
Code seems :100: but why are we scoping this in the first place? 

I also don't see this documented anywhere (like here: https://www.notion.so/cypressdx/Changes-for-10-0-0-1dca213837b84d6ea297c2d12c615f3a#d80e7d92e76c41b9b56673acf48f93cf).

Sorry to drop this here - just wanting to clarify whether 

1. DX decided this and forgot to document it
2. Someone else decided this without it going through the correct process (usually DX should approve config changes)

cc @cowboy perhaps, master of the config changes document.

**Review by jennifer-shehane — DISMISSED** [review]
@BlueWinds Can you open a PR against the unification docs? The slowTestThreshold needs to be moved into the Testing-Type Specific Options for config: https://deploy-preview-4186--cypress-docs.netlify.app/guides/references/configuration#Testing-Type-Specific-Options

**Comment by BlueWinds** [comment]
> @BlueWinds Can you open a PR against the unification docs? The slowTestThreshold needs to be moved into the Testing-Type Specific Options for config: https://deploy-preview-4186--cypress-docs.netlify.app/guides/references/configuration#Testing-Type-Specific-Options

Done in https://github.com/cypress-io/cypress-documentation/pull/4388. Was on my TODO list, is why I didn't check it off in the PR Tasks checklist. :)

File: packages/config/lib/options.ts (hunk 2 above)
Comment (lmiller1990) [reply packages/config/lib/options.ts:569]: Update: we decided it's going to non root level. I still find this a little weird, but I don't think it's incorrect.

File: packages/data-context/test/unit/sources/migration/codegen.spec.ts [hunk 3]
```diff
@@ -42,7 +42,7 @@ describe('cypress.config.js generation', () => {
...
```
Comment (lmiller1990) [inline packages/data-context/test/unit/sources/migration/codegen.spec.ts:45]: @BlueWinds  I think the logic here is "If you have not configured component testing (most users from 9.x are E2E only) then we will **not** add the `component: { setupNodeEvents }` key to your migrated configuration. The presence of this key is used to determine if you've configured component testing, I think.

I'd say we want at least one test that shows when `hasComponentTesting: false`, we are not populating the `component` key in the migrated configuration file during migration.

**Review by lmiller1990 — COMMENTED** [review]
One comment! We don't want to add `component` to the migrated config if you haven't previously used component testing; we should have a test where `hasComponentTesting: false`, which means "this user has not used component testing before".

File: packages/data-context/test/unit/sources/migration/codegen.spec.ts [hunk 4]
```diff
@@ -58,6 +58,34 @@ describe('cypress.config.js generation', () => {
...
+      hasPluginsFile: true,
+      projectRoot,
+      hasTypescript: false,
+    })
+
+    snapshot(generatedConfig)
+  })
+
+  it('should create only a component entry when no e2e specs are detected', async () => {
```
Comment (BlueWinds) [inline packages/data-context/test/unit/sources/migration/codegen.spec.ts:72]: @lmiller1990 - I added this test and left the one below it, testing the case where only one testing type is detected. The snapshots for these two test display only `component` and only `e2e` blocks respectively.

File: packages/data-context/test/unit/sources/migration/codegen.spec.ts (hunk 3 above)
Comment (BlueWinds) [reply packages/data-context/test/unit/sources/migration/codegen.spec.ts:45]: I already added a test which demonstrates that, with a snapshot. @'ed you on that line.

File: packages/config/lib/options.ts [hunk 5]
```diff
@@ -565,6 +565,12 @@ export const breakingRootOptions: Array<BreakingOption> = [
...
+    errorKey: 'CONFIG_FILE_INVALID_ROOT_CONFIG',
+    isWarning: false,
+    testingTypes: ['component', 'e2e'],
+  },
```
Comment (emilyrohrbough) [inline packages/config/lib/options.ts:573]: On this PR, @estrada9166 added/updated typescript tests for baseUrl, which is another invalid root-level configuration option. Might be a good idea to add in tests for this option. Thoughts?

https://github.com/cypress-io/cypress/pull/20589/files#diff-facdf706a5238380b5e97c040bc1a5f5cb89e3116f878754dd7fb4e693787715L35

File: packages/data-context/src/sources/migration/codegen.ts [hunk 6]
```diff
@@ -340,91 +340,86 @@ export function renameSupportFilePath (relative: string) {
...
+          return acc
+        }
+
+        const { testFiles, ignoreTestFiles, ...rest } = value
+
+        // don't include if it's the default! No need.
+        const specPattern = getSpecPattern(cfg, key)
+        const ext = '**/*.cy.{js,jsx,ts,tsx}'
+        const isDefaultE2E = key === 'e2e' && specPattern === `cypress/e2e/${ext}`
+        const isDefaultCT = key === 'component' && specPattern === ext
```
Comment (emilyrohrbough) [inline packages/data-context/src/sources/migration/codegen.ts:363]: not a change for this PR, just in general..would love to see these default values pulled from the config packages instead of hard-coded here

File: packages/data-context/src/sources/migration/codegen.ts [hunk 7]
```diff
@@ -340,91 +340,86 @@ export function renameSupportFilePath (relative: string) {
...
+      case 'ignoreTestFiles':
+        return {
+          ...acc,
+          e2e: { ...acc.e2e, specExcludePattern: val },
+          component: { ...acc.component, specExcludePattern: val },
+        }
+      case 'supportFile':
+        return {
+          ...acc,
+          e2e: { ...acc.e2e, supportFile: val },
```
Comment (emilyrohrbough) [inline packages/data-context/src/sources/migration/codegen.ts:407]: hmm sorry @BlueWinds...sanity check though, currently in 9x the support file is shared between e2e and component right? Shouldn't this map to both?

File: packages/data-context/test/unit/sources/migration/codegen.spec.ts (hunk 1 above)
Comment (emilyrohrbough) [reply packages/data-context/test/unit/sources/migration/codegen.spec.ts:27]: I'm curious as well why this changed.

File: packages/config/lib/options.ts (hunk 5 above)
Comment (BlueWinds) [reply packages/config/lib/options.ts:573]: Looking at those tests, it doesn't look like we test every config value that way, just a few examples selected to show `Cypress.config()` reflects the typings of values that can be retrieved from it. We already have `Cypress.config('taskTimeout') // $ExpectType number` demonstrating that numbers are done correctly, and don't think slowTestThreshold needs to be added there.

Interesting though, I'd never seen that file before.

File: packages/data-context/src/sources/migration/codegen.ts [hunk 8]
```diff
@@ -340,91 +340,86 @@ export function renameSupportFilePath (relative: string) {
...
+        return {
+          ...acc,
+          e2e: { ...acc.e2e, specPattern: getSpecPattern(cfg, 'e2e') },
+          component: { ...acc.component, specPattern: getSpecPattern(cfg, 'component') },
+        }
+      case 'ignoreTestFiles':
+        return {
+          ...acc,
+          e2e: { ...acc.e2e, specExcludePattern: val },
+          component: { ...acc.component, specExcludePattern: val },
```
[the remark this answers is not in the document]
Comment (BlueWinds) [reply packages/data-context/src/sources/migration/codegen.ts:402]: Thanks for opening an issue to track that.

File: packages/data-context/src/sources/migration/codegen.ts [hunk 9]
```diff
@@ -340,91 +340,86 @@ export function renameSupportFilePath (relative: string) {
...
+          e2e: { ...acc.e2e, [key]: val },
+        }
+      case 'slowTestThreshold':
+        return {
+          ...acc,
+          component: { ...acc.component, [key]: val },
+          e2e: { ...acc.e2e, [key]: val },
+        }
+      default:
+        return { ...acc, global: { ...acc.global, [key]: val } }
```
[the remark this answers is not in the document]
Comment (BlueWinds) [reply packages/data-context/src/sources/migration/codegen.ts:421]: `switch` took me a while to warm up to, but after working with Redux (where it's a standard pattern in reducers, which you write a lot of) it feels super comfy.

File: packages/data-context/src/sources/migration/codegen.ts (hunk 7 above)
Comment (BlueWinds) [reply packages/data-context/src/sources/migration/codegen.ts:407]: Hm, it does sound like it should to me. @lmiller1990, thoughts?

File: packages/data-context/test/unit/sources/migration/codegen.spec.ts (hunk 1 above)
Comment (BlueWinds) [reply packages/data-context/test/unit/sources/migration/codegen.spec.ts:27]: `should create only an e2e entry when no component specs are detected`, added below, asserts that `component` is left out when `hasComponenttesting: false`, so that case is still covered.

Previously we didn't have any tests covering both CT and e2e testing both being true. With this switched to true, we're asserting

"this global option remains global and doesn't end up in any testing-type specific config"
rather than
"this global option remains global and doesn't end up in the e2e config"

I think the default should be asserting on the shape of both properties, and that a separate assertion "we leave out e2e/ct when only one testing type is detected" (which I added below) is the way to go.

File: packages/data-context/src/sources/migration/codegen.ts (hunk 7 above)
Comment (lmiller1990) [reply packages/data-context/src/sources/migration/codegen.ts:407]: I asked exactly the same question at the time - Brian said nope, we just map to `e2e.supportFile`. During CT setup we will create a new file called `cypress/support/component.js`.

(Yes - this means that CT users will have some small inconvenience during migration from 9.x -> 10.x. But CT users are in the minority).

You can ping in the migration channel if you feel this is weird and should be revisited.

File: packages/data-context/test/unit/sources/migration/codegen.spec.ts (hunk 4 above)
Comment (lmiller1990) [reply packages/data-context/test/unit/sources/migration/codegen.spec.ts:72]: Thanks!

**Review by lmiller1990 — APPROVED** [review]
@emilyrohrbough pointed out lots of potential small improvements we could make, I don't see those as blockers to this PR, since those were not changes introduced here. Definitely good to capture those in Jira tickets.

As for the actual code and PR, my comments have been addressed. ✅

File: packages/data-context/src/sources/migration/codegen.ts (hunk 7 above)
Comment (emilyrohrbough) [reply packages/data-context/src/sources/migration/codegen.ts:407]: @lmiller1990 I did ask there as well. No response yet.

**Review by emilyrohrbough — APPROVED** [review]

File: packages/config/lib/options.ts (hunk 5 above)
Comment (emilyrohrbough) [reply packages/config/lib/options.ts]: It looks like it can be added to verify the types are working as expected and/or throws an error when used incorrectly. i.e. 

```js
// cypress.config.js
{
   slowTestThreshold: 200 // expect error as unsupported type
}```

Comment (BlueWinds) [reply packages/config/lib/options.ts]: `Cypress.config('slowTestThreshold', 1000)` is not an error, nor should it be - in the resolved config, `slowTestThreshold` can be at the top level, and it can be changed on a per-test basis (unlike baseUrl). It's only in the config file that it can't be top-level.
